docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,507 @@
+With the introduction of ClusterClass, we are extending the repsonsibility of the Cluster controller to create/update the actual objects from the collection of templates. Following this pattern, a template is required to create the actual infrastructure cluster object as part of the managed topology. For starter, the template should consist the InfraClusterSpec object. It can be enhanced to include other fields such as labels/annotations in the future.
+
+The proposal calls for the implementation of an infrastructure cluster template to be referenced by the `ClusterClass` object. This is a required field for the `ClusterClass` object. This template would be used to create an infrastructure cluster object. Currently, this responsibility is assumed by the cluster operator/end user.
+
+##### For Control plane providers
+A control plane provider should create a template for use of creation for the control plane objects. For instance, the kubeadm control plane provider should introduce the notion of a KubeadmControlPlaneTemplate for use by the cluster controller to create a managed Kubeadm Control plane object. The minimum set of parameters supported by the control plane template are:
+  - infrastructure template reference
+
+### Risks and Mitigations
+
+This proposal tries to model the API design for ClusterClass with a narrow set of use cases. This initial implementation provides a baseline on which incremental changes can be introduced in the future. Instead of encompassing of all use cases under a single proposal, this proposal mitigates the risk of waiting too long to consider all required use cases under this topic.
+
+## Alternatives
+
+The ClusterClass approach is to ease the UX of cluster creation. This is an alternative way of easily creating multiple clusters of similar flavor.
+CAPI still supports the default cluster creation functionality wherein the users can supply the individual object definitions (KCP, MD, etc.) which make up a cluster.
+
+## Upgrade Strategy
+
+Existing clusters created without ClusterClass cannot switch over to using ClusterClass for a managed topology.
+
+## Additional Details
+
+### Test Plan [optional]
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.
+Anything that would count as tricky in the implementation and anything particularly challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage expectations).
+Please adhere to the [Kubernetes testing guidelines][testing-guidelines] when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+
+### Graduation Criteria [optional]
```

Comment:
Is this intending to go under a gate flag initially?

Reply:
The intention is to ship it under a feature gate initially. Will add that to the proposal.